Add/remove more than one plugin at once

joshpurvis opened the pull request:
Allow `cs plugin add` and `cs plugin remove` commands to operate on multiple plugins at once:

    cs plugin add cloud cloud-hints navigator

normanjoyner commented:
@joshpurvis can we have this install the plugins regardless of whether or not they are found in `authorized_plugins`?  Users should be able to install from github repo urls or via npm packages.

joshpurvis commented:
Refactored to install/remove plugins regardless if they're in the `authorized_plugins` list.

I couldn't find any documentation on `npm.commands.install`, but it appears to be an asynchronous method. Is there any harm in the way I'm relying on the fact that there's nothing else in the stack here? 

Or should I introduce something like `async.eachSeries` to force an order, and control the flow better?

normanjoyner commented:
I think we will need have the `npm.commands.install` commands within an `async.each`, and only install one plugin at a time so if one fails, it doesn't prevent others from being installed.

joshpurvis commented:
Ok I updated it to use `async.eachSeries`, because I'm assuming that's what you meant by saying this:

> and only install one plugin at a time so if one fails, it doesn't prevent others from being installed.

I believe using `async.each` would have behaved almost exactly as I had it in the previous commit.. which I just rebased, but it roughly looked like:

```
_.each(option.plugin, function(plugin) {
   // ... 
   npm.command.install ( ... );  // this is async, but even if one failed, 
                                 // it still installed the others.
});
```

Sorry for all the confusion.. just trying to better understand best practices for async patterns in JS.

normanjoyner commented on commands/plugin.js:
```diff
@@ -148,17 +149,26 @@ module.exports = {
+
+                            npm.commands.install([plugin], function(err, data) {
+                                if (err)
+                                    console.log(["Failed to install plugin:", plugin].join(" "))
```
Can this be changed to `process.stderr.write` rather than `console.log`? A line break will have to be manually added to the end of the line as well as `process.stderr.write` does not include one by default, as `console.log` does.

normanjoyner commented on commands/plugin.js:
```diff
@@ -186,21 +196,30 @@ module.exports = {
+                            } 
+                            npm.commands.uninstall([plugin], function(err, data){
+                                if (err)
+                                    console.log(["Failed to uninstall", plugin].join(" "));
```
Can this be changed to `process.stderr.write` rather than `console.log`? A line break will have to be manually added to the end of the line as well as `process.stderr.write` does not include one by default, as `console.log` does.

normanjoyner commented on commands/plugin.js:
```diff
@@ -186,21 +196,30 @@ module.exports = {
+                            npm.commands.uninstall([plugin], function(err, data){
+                                if (err)
+                                    console.log(["Failed to uninstall", plugin].join(" "));
+                                cb();
```
while not necessary here, it is a good practice to always `return cb();`

normanjoyner commented on commands/plugin.js:
```diff
@@ -148,17 +149,26 @@ module.exports = {
+                            npm.commands.install([plugin], function(err, data) {
+                                if (err)
+                                    console.log(["Failed to install plugin:", plugin].join(" "))
+                                cb();
```
while not necessary here, it is a good practice to always `return cb();`

normanjoyner commented on commands/plugin.js:
```diff
@@ -148,17 +149,26 @@ module.exports = {
-                        console.log(["Installing plugin:", options.plugin].join(" "));
-                        npm.commands.install([options.plugin], function(err, data){});
+                    }, function () {
+                        async.eachSeries(options.plugin, function(plugin, cb) {
```
This can be changed to `async.each` which will start all the plugin installs before waiting for the previous install to complete. Utilizing `async.each` over `_.each` will allow for a callback to be executed when the loop is finished.

normanjoyner commented on commands/plugin.js:
```diff
@@ -186,21 +196,30 @@ module.exports = {
-                            if(options.plugin.indexOf(".git") != -1)
-                                options.plugin = options.plugin.substring(0, options.plugin.indexOf(".git"));
-                        }
+                        async.eachSeries(options.plugin, function(plugin, cb) {
```
This can be changed to `async.each` which will start all the plugin installs before waiting for the previous install to complete. Utilizing `async.each` over `_.each` will allow for a callback to be executed when the loop is finished.

joshpurvis commented:
> Utilizing async.each over _.each will allow for a callback to be executed when the loop is finished.

Thanks, this explanation helps. I wasn't sure of the difference, as simply testing their behaviour produced the same output. (I think because `npm.commands.install` is also an asynchronous method)

> Can this be changed to process.stderr.write rather than console.log?

Should the console.logs be `stdout` instead?

Anyway, made the `async` and other requested changes. Glad to see you've added CI, by the way. Which testing framework do you prefer?

normanjoyner commented:
@joshpurvis leaving the `console.log` is fine, as it writes to stdout. I've been using mocha for testing.
